BDArmory/Modules/ModuleWeapon.cs

```diff
@@ -799,17 +827,33 @@ void FixedUpdate()
                 }
 
 
-                if (vessel.isActiveVessel)
-                {
-                    if (showReloadMeter)
-                    {
-                        UpdateReloadMeter();
-                    }
-                    else
-                    {
-                        UpdateHeatMeter();
-                    }
-                }
+                if (vessel.isActiveVessel)// lets not have it re-initialize and reenumerate gauges and ammo counts every tick. Seriously, what was I thinking?
+		{
+			if (showReloadMeter)
+			{
+				UpdateReloadMeter();
+			}
+			else
+			{
+				UpdateHeatMeter();
+			}
+			UpdateAmmo();
+			if (!isOutofAmmo && ammoGauge == null) //only redraw these if nulled from vessel switch
+			{
+				UpdateAmmoMeter();
+			}
+			if (isOutofAmmo && emptyGauge == null)
+			{
+				UpdateEmptyAlert();
+			}
+			updateList -= Time.fixedDeltaTime;
+			if (updateList < 0)
+			{
+				UpdateAmmoMeter(); //call once a second to check for infinite ammo toggle, sanity check, etc
+				UpdateEmptyAlert();
+				updateList = 1.0f; 
+			}
+		}
```

Comment:
I know it was done here before this PR, but doing UI in `FixedUpdate` is wrong, it should all go into `Update`. Unity tries to run `FixedUpdate` 50 times a second whatever happens, and `Update` as much as it can (i.e. once every frame). This means in heavy load situations it will try to update the GUI 50 times a second even though the game is chugging along at only 10 FPS.

Premature optimization is the root of a lot of evil. :) But it would still be tidier to have it in update, and we already know the gauges have a performance impact.

Reply:
Ah. that would explain it. Gauges got put into FixedUpdate since that's where the heat and reload gauge init code was, but if it's being called 50X more often than it needs to be, no wonder the initial version with a per-weapon ammocount enum caused lag. In that case, lines 832-859 should all get moved to Update()